pkg/ntpmonitor/ntpmonitor.go

```diff
@@ -0,0 +1,151 @@
+//
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ntpmonitor
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/beevik/ntp"
+
+	"github.com/sigstore/timestamp-authority/pkg/log"
+)
+
+var (
+	// ErrInvTime indicates that the local time has drifted too much
+	// from the monitored NTP servers.
+	ErrInvTime = errors.New("local time differs from observed")
+	// ErrTooFewServers means that the number of trusted servers are
+	// smaller then the selected num servers to query.
+	ErrTooFewServers = errors.New("too few ntp servers configured")
+	// ErrNoResponse indicates that there is an error to communicate with
+	// the remote NTP servers
+	ErrNoResponse = errors.New("no ntp response")
+)
+
+// NTPMonitor compares the local time with a set of trusted NTP servers.
+type NTPMonitor struct {
+	cfg *Config
+	run atomic.Bool
+}
+
+// New creates a NTPMonitor, reading the configuration from the provided
+// path.
+func New(configFile string) (*NTPMonitor, error) {
+	cfg, err := LoadConfig(configFile)
+	if err != nil {
+		return nil, err
+	}
+	return NewFromConfig(cfg), nil
+}
+
+// NewFromConfig creates a NTPMonitor from an instantiated configuration.
+func NewFromConfig(cfg *Config) *NTPMonitor {
+	return &NTPMonitor{cfg: cfg}
+}
+
+// Start the periodic monitor. If there is an initialization error, the
+// function returns immediately. Once the periodic monitoring starts, it runs
+// until an error is found, or Stop() is being called.
+func (n *NTPMonitor) Start() error {
+	n.run.Store(true)
+
+	if len(n.cfg.Servers) < n.cfg.NumServers {
```

Comment:
Can configuration errors be checked on initialization (New()) rather than in Start()?

Reply:
Updated.